pkg/httpproxy/proxy.go

```diff
@@ -229,27 +234,96 @@ func (p *proxy) secretGetter(req *http.Request, cAuth string) SecretGetter {
 	}
 }
 
-func (p *proxy) checkCluster(req *http.Request, user user.Info, clusterID, credID string) (authorizer.Decision, error) {
-	cluster, err := p.clusters.Controller().Lister().Get("", clusterID)
-	if err != nil {
+// checkIndirectAccessViaCluster checks if the user has access to the cloud credential via being owner of a cluster associated to the cloud credential.
+// Currently, only EKS and provisioningv2 clusters are supported because those clusters have a cloud credential associated to them.
+// GKE and AKS clusters also have cloud credential associated to them, but those are checked via specific proxies (not the meta proxy).
+func (p *proxy) checkIndirectAccessViaCluster(req *http.Request, user user.Info, clusterID, credID string) (authorizer.Decision, error) {
+	var (
+		mgmtClusters []*v3.Cluster
+		provClusters []*prov.Cluster
+		err          error
+
+		i int
+		c interface{}
+	)
+	if clusterID == "" {
+		// If no clusterID is passed, then we check all clusters that the user has access to and are associated to the cloud credential.
+		// Both management and provisioning clusters should be checked.
+		mgmtClusters, err = p.mgmtClustersCache.GetByIndex(cluster.ByCloudCredential, credID)
+		if err != nil {
+			return authorizer.DecisionDeny, err
+		}
+
+		provClusters, err = p.provClustersCache.GetByIndex(provcluster.ByCloudCred, credID)
+	} else {
+		var c *v3.Cluster
+		c, err = p.mgmtClustersCache.Get(clusterID)
+		mgmtClusters = []*v3.Cluster{c}
+	}
+	if err != nil || len(mgmtClusters)+len(provClusters) == 0 {
 		return authorizer.DecisionDeny, err
 	}
-	if cluster.Spec.EKSConfig == nil {
-		return authorizer.DecisionDeny, nil
+
+	c = nextCluster(mgmtClusters, provClusters, &i)
+	for c != nil {
+		switch cluster := c.(type) {
+		case *v3.Cluster:
+			administratedByProvV2Cluster := imported.IsAdministratedByProvisioningCluster(cluster)
+			if (cluster.Spec.EKSConfig == nil || cluster.Spec.EKSConfig.AmazonCredentialSecret != credID) && !administratedByProvV2Cluster {
+				continue
+			}
+
+			if administratedByProvV2Cluster {
+				c, err = p.provClustersCache.GetByIndex(provcluster.ByCluster, cluster.Name)
+				if err != nil {
+					c = nextCluster(mgmtClusters, provClusters, &i)
+				}
+				// If there is no error, then we want to go through the loop again with the provisioning cluster.
```

Comment:
I suppose this is fine but if there is a provisioning cluster won't it hit it eventually anyway? I think just allowing that to happen would simplify the logic. I think it would also make it to where nextCluster could just be used in the beginning of the loop.

Reply:
The way is written it won't find the provisioning cluster if the `clusterID` is provided. I will update the logic so that this will no longer be the case.